### ISSUE TITLE
[Potential Bug] performance difference is quite significant for GPU and CPU schemes under horizontal federated learning

### ISSUE THREAD

**Issue by ZiyueXu77** [dmlc/xgboost#10612]
The two training schemes for horizontal federated learning - CPU and GPU, produce different models with quite different results w.r.t. evaluation accuracy. Potentially one of them does not align with the algorithm assumptions.

With the data, data split, and training scripts used in [this example](https://github.com/ZiyueXu77/NVFlare/tree/secureboost/examples/advanced/xgboost_secure), the two schemes (`train_hori_base.py` v.s. `train_hori_base_gpu.py`) have the following training performance based on the same training data:

Horizontal federated with CPU
```
[0]	eval-auc:0.89789	train-auc:0.92732
[1]	eval-auc:0.89791	train-auc:0.92733
[2]	eval-auc:0.89791	train-auc:0.92733
```
Horizontal federated with GPU
```
[0]	eval-auc:0.90514	train-auc:0.92730
[1]	eval-auc:0.90515	train-auc:0.92731
[2]	eval-auc:0.90517	train-auc:0.92732
```

For reference, the baseline centralize training result (`train_base.py` and `train_base_gpu` almost the same results) is 
```
[0]	eval-auc:0.90515	train-auc:0.92747
[1]	eval-auc:0.90517	train-auc:0.92748
[2]	eval-auc:0.90518	train-auc:0.92748
```

This may not be a bug that breaks the pipeline, but I think we may want to find out the root cause.

**Comment by trivialfis** [comment]
The result looks normal, can you run a few more rounds, like 32? The GPU can CPU are expected to reproduce different results.

**Comment by ZiyueXu77** [comment]
I see, yeah I run with 50 rounds, and the discrepancies shrink after a few rounds.
The results at the end of 50 rounds are below:
```
centralized:            eval-auc:0.94752	train-auc:0.98835
centralized_gpu:   eval-auc:0.94812	train-auc:0.98840
hori:                        eval-auc:0.94707	train-auc:0.98883
hori_gpu:                eval-auc:0.94761	train-auc:0.98875
```
The results can be considered within expected range, closing this issue
